app/services/topology_service.rb

```diff
@@ -28,15 +28,42 @@ def build_base_entity_data(entity)
     {
       :name   => entity.name,
       :kind   => entity_type(entity),
-      :miq_id => entity.id
+      :model  => entity.class.to_s,
+      :miq_id => entity.id,
+      :key    => entity_id(entity)
     }
   end
 
+  def group_nodes_by_model(nodes)
+    if block_given?
+      nodes_groupped_by_model = nodes.group_by { |_, v| v[:model] }
+
+      nodes_groupped_by_model.each do |klass, entity_data|
+        yield(klass, entity_data.map { |x| [x.second[:miq_id], x.second[:key]] }.to_h)
+      end
+    end
+  end
+
   def build_topology
     included_relations = self.class.instance_variable_get(:@included_relations)
     preloaded = @providers.includes(included_relations)
     nodes, edges = map_to_graph(preloaded, build_entity_relationships(included_relations))
 
+    remove_list = []
```

Comment:
This can be a method. e.g. `disallowed_nodes`.

Then the below can be a method e.g. `filter_nodes_and_edges`

Then then
```
filtered_edges, filtered_nodes = filter_nodes_and_edges(nodes, edges, disallowed_nodes...
```

please, figure out the details, but this method is pretty long

Thx!